hw/xbox/mcpx_apu.c

```diff
@@ -473,6 +514,120 @@ static void ep_scratch_rw(void *opaque,
                       ptr, addr, len, dir);
 }
 
+static uint32_t circular_scatter_gather_rw(MCPXAPUState *d,
+                                           hwaddr sge_base,
+                                           unsigned int max_sge,
+                                           uint8_t *ptr,
+                                           uint32_t base, uint32_t end,
+                                           uint32_t cur,
+                                           size_t len,
+                                           bool dir)
+{
+    while (len > 0) {
+        unsigned int bytes_to_copy = end - cur;
+
+        if (bytes_to_copy > len) {
+            bytes_to_copy = len;
+        }
+
+        MCPX_DPRINTF("circular scatter gather %s in range 0x%x - 0x%x at 0x%x of length 0x%x / 0x%lx bytes\n",
+            dir ? "write" : "read", base, end, cur, bytes_to_copy, len);
+
+        assert((cur >= base) && ((cur + bytes_to_copy) <= end));
+        scatter_gather_rw(d, sge_base, max_sge, ptr, cur, bytes_to_copy, dir);
+
+        ptr += bytes_to_copy;
+        len -= bytes_to_copy;
+
+        /* After the first iteration we might have to wrap */
+        cur += bytes_to_copy;
+        if (cur >= end) {
+            assert(cur == end);
+            cur = base;
+        }
+    }
+
+    return cur;
+}
+
+static void gp_fifo_rw(void *opaque, uint8_t *ptr,
+                       unsigned int index, size_t len,
+                       bool dir)
+{
+    MCPXAPUState *d = opaque;
+    uint32_t base;
+    uint32_t end;
+    hwaddr cur_reg;
+    if (dir) {
+        assert(index < GP_OUTPUT_FIFO_COUNT);
+        base = GET_MASK(d->regs[NV_PAPU_GPOFBASE0 + 0x10 * index],
+                        NV_PAPU_GPOFBASE0_VALUE);
+        end = GET_MASK(d->regs[NV_PAPU_GPOFEND0 + 0x10 * index],
+                       NV_PAPU_GPOFEND0_VALUE);
+        cur_reg = NV_PAPU_GPOFCUR0 + 0x10 * index;
+    } else {
+        assert(index < GP_INPUT_FIFO_COUNT);
+        base = GET_MASK(d->regs[NV_PAPU_GPIFBASE0 + 0x10 * index],
+                        NV_PAPU_GPOFBASE0_VALUE);
+        end = GET_MASK(d->regs[NV_PAPU_GPIFEND0 + 0x10 * index],
+                       NV_PAPU_GPOFEND0_VALUE);
+        cur_reg = NV_PAPU_GPIFCUR0 + 0x10 * index;
+    }
+
+    uint32_t cur = GET_MASK(d->regs[cur_reg], NV_PAPU_GPOFCUR0_VALUE);
+
+    /* DSP hangs if current >= end; but forces current >= base */
+    assert(cur < end);
+    if (cur < base) {
+        cur = base;
+    }
+
+    cur = circular_scatter_gather_rw(d,
+        d->regs[NV_PAPU_GPFADDR], d->regs[NV_PAPU_GPFMAXSGE],
+        ptr, base, end, cur, len, dir);
+
+    SET_MASK(d->regs[cur_reg], NV_PAPU_GPOFCUR0_VALUE, cur);
+}
+
+static void ep_fifo_rw(void *opaque, uint8_t *ptr,
```

Comment:
This function appears to be identical to `gp_fifo_rw`, except for the register offsets. We may want to refactor this into an abstract function. I wouldn't block merge as-is, but food for thought if we develop on it further.

Reply:
Yes, this is one of the key design-questions I have had: https://github.com/xqemu/xqemu/pull/167#discussion_r244278110

Reply:
[ Replying to above comment... ]